File: data-structures/javascript/merged-sorted-arrays.js

// Made by me
function mergeSortedArrays1(arr1, arr2) {
    let mergedSortedArray = [];

    if (arr1.length === 0) {
        return arr2;
    }
    if (arr2.length === 0) {
        return arr1;
    }

    let index1 = 0;
    let index2 = 0;

    while (index1 !== arr1.length && index2 !== arr2.length) {
        console.log(index1, ' - ', index2);
        const value1 = arr1[index1];
        const value2 = arr2[index2];

        if (value1 <= value2) {
            mergedSortedArray.push(value1);
            index1++;
        } else if (value1 > value2) {
            mergedSortedArray.push(value2);
            index2++;
        }
    }
    console.log(index1, ' - ', index2);
    const leftOver1 = arr1.slice(index1);
    const leftOver2 = arr2.slice(index2);

    mergedSortedArray = [...mergedSortedArray, ...leftOver1, ...leftOver2];

    return mergedSortedArray;
}


// Made by the teacher
function mergeSortedArrays(array1, array2){
  const mergedArray = [];

  if(array1.length === 0) {
    return array2;
  }
  if(array2.length === 0) {
    return array1;
  }

  let array1Item = array1[0];
  let array2Item = array2[0];
  let i = 1;
  let j = 1;

  while (array1Item || array2Item){
   if(array2Item === undefined || array1Item < array2Item){
     mergedArray.push(array1Item);
     array1Item = array1[i];
     i++;
   }   
   else {
     mergedArray.push(array2Item);
     array2Item = array2[j];
     j++;
   }
  }
  return mergedArray;
}


console.log(mergeSortedArrays([0,3,4,31, 65], [4,6,30,70]));

// index1 = 3
// index2 = 2
// mergedLength = 7

File: data-structures/python/merged-sorted-arrays.py

a = [0,3,4,31, 65]
b = [4,6,30,70]

def mergeSortedArrays(arr1, arr2):
  point1 = 0
  point2 = 0
  mergedArray = []

  if len(arr1) == 0:
    return arr2
  if len(arr2) == 0:
    return arr1

  while not point1 == len(arr1) or not point2 == len(arr2):
    if point1 == len(arr1):
      mergedArray.append(arr2[point2])
      point2 += 1
      continue
    if point2 == len(arr2):
      mergedArray.append(arr1[point1])
      point1 += 1
      continue

    value1 = arr1[point1]
    value2 = arr2[point2]
    if value1 == value2:
      mergedArray.append(value1)
      mergedArray.append(value2)
      point1 += 1
      point2 += 1
    elif value1 < value2:
      mergedArray.append(value1)
      point1 += 1
    else:
      mergedArray.append(value2)
      point2 += 1

  return mergedArray 
    


print(mergeSortedArrays(a, b))
